test/e2e/cluster_upgrade_test.go

```diff
@@ -69,9 +83,117 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
+						return 0, errors.New("failed to list kube-proxy pods in workload cluster")
+					}
+
+					errList := []error{}
+
+					// Check all nodes to be Ready.
+					for _, node := range nodes.Items {
+						for _, condition := range node.Status.Conditions {
+							if condition.Type != corev1.NodeReady || condition.Status == corev1.ConditionTrue {
+								continue
+							}
+							errList = append(errList, errors.Errorf("Node's %s Ready condition is false", node.GetName()))
+						}
+					}
+
+					// Check if the expected number of kube-proxy pods exist and all of them are healthy for all existing Nodes of the Cluster.
+					if len(nodes.Items) != len(kubeProxyPods.Items) {
+						errList = append(errList, errors.Errorf("exected %d kube-proxy pods to exist, got %d", len(nodes.Items), len(kubeProxyPods.Items)))
+					}
+					for _, pod := range kubeProxyPods.Items {
+						for _, condition := range pod.Status.Conditions {
+							if condition.Type != corev1.PodReady || condition.Status == corev1.ConditionTrue {
+								continue
+							}
+							errList = append(errList, errors.Errorf("Pod's %s Ready condition is false", pod.GetName()))
+						}
+					}
+
+					if err := kerrors.NewAggregate(errList); err != nil {
+						return 0, errors.Wrap(err, "blocking upgrade because cluster is not stable")
+					}
+
+					// At this stage all current machines are considered ok, so remove the pre-drain webhook from a CP to unblock the next step of the upgrade.
+					if len(deletingMachinesWithPreDrainHook) > 0 {
+						if len(deletingMachinesWithPreDrainHook) > 1 {
+							return 0, errors.Errorf("expected a maximum of 1 control-plane machines to be in deleting and having the pre-drain hook but got %d", len(deletingMachinesWithPreDrainHook))
+						}
+
+						// Removing pre-drain hook from machine.
+						m := &deletingMachinesWithPreDrainHook[0]
+						patchHelper, err := patch.NewHelper(m, managementClusterProxy.GetClient())
+						if err != nil {
+							return 0, errors.Wrapf(err, "creating patchHelper for Machine %s", klog.KObj(m))
+						}
+						delete(m.Annotations, preDrainHook)
+
+						if err := patchHelper.Patch(ctx, m); err != nil {
+							return 0, errors.Wrapf(err, "failed to remove pre-drain annotation from Machine %s", klog.KObj(m))
+						}
+
+						// Return to enter the function again.
+						return 0, errors.Errorf("deletion of Machine %s was blocked by pre-drain hook", klog.KObj(m))
+					}
+
+					if int64(len(machines)) > upgraded {
+						return 0, errors.New("old Machines remain")
+					}
+
+					return upgraded, nil
+				}, e2eConfig.GetIntervals("k8s-upgrade-and-conformance", "wait-machine-upgrade")...).Should(Equal(controlPlaneMachineCount), "Timed out waiting for all control-plane machines in Cluster %s to be upgraded to kubernetes version %s", klog.KObj(cluster), cluster.Spec.Topology.Version)
```

Comment:
Does it work correctly to check for only a number while returning a number and an error?

Reply:
I think I would just pass in g Gomega into this eventually and then check everything (and not return anything)

Reply:
Hm but not sure if that works with the retry logic you need

Reply:
its the same pattern we already use in WaitForControlPlaneMachinesToBeUpgraded

Reply:
```
If multiple values are returned by the function, Eventually will pass the first value to the matcher and require that all others are zero-valued.
```
From the godoc

Reply:
Alright, sounds good!

Reply:
Didn't remember